internal/deb/extract.go

```diff
@@ -268,3 +295,21 @@ func extractData(dataReader io.Reader, options *ExtractOptions) error {
 
 	return nil
 }
+
+func orderedParentDirs(path string) []string {
+	var parents []string
+	path = filepath.Clean(path)
+	for {
+		path = filepath.Dir(path)
+		if path == "/" {
+			break
+		}
+		parents = append(parents, path+"/")
+	}
+
+	// Reverse the list.
+	for i := 0; i < len(parents)/2; i++ {
+		parents[i], parents[len(parents)-i-1] = parents[len(parents)-i-1], parents[i]
```

Comment:
This function can be made both simpler and significantly faster by reusing the string we already got, which means no further allocations. We can also allocate an array of the correct length by counting slashes.

For example:
```
func parentDirs(path string) []string {
	parents := make([]string, strings.Count(path, "/"))
	count := 0
	for i, c := range path {
		if c == '/' {
			parents[count] = path[:i+1]
			count++
		}
	}
	return parents
}
```

https://go.dev/play/p/DKAgL045JYu

Reply:
That is indeed much better, thanks! I was not thinking about the extra allocation required by `+ '/'`. The only problem I see with the solution is that it will not handle `utf-8` because of the `i+1` where we really want to stop at the next rune, not at the next byte.

Reply:
I have changed the implementation to handle uft-8 making it slightly more complex by looping manually (see https://go.dev/blog/strings#libraries). I have also added a test case to make sure we handle utf-8 at the extractor level.

Reply:
How's the original code mishandling a UTF-8 string due to the +1?  When is "/" not one byte long?

Reply:
You are right :D, I saw the plus one on the byte count for utf-8 and I automatically changed it to use the character width, I am reverting it back